### PR TITLE
Introduce `://` syntax for Windows Devices in DeviceMapping.PathOnHost

### PR DESCRIPTION
**Pull request by TBBle**
**- What I did**

Introduced a new Windows Device string syntax for the `--devices` and the `api.types.container.DeviceMapping.PathOnHost`.

The syntax is `IDType://ID`, reflecting an evolution and generalisation of the existing `class/ID` syntax. The latter remains supported specifically for `class/X`: `IDType/ID` for other `IDType`s is not implemented to avoid future parsing ambiguity.

This syntax will function with both the `--device` parameter from https://github.com/docker/cli, and also when fed from Kubernetes (or other CRI client) via https://github.com/Mirantis/cri-dockerd, or [dockershim](https://kubernetes.io/blog/2022/02/17/dockershim-faq/) where supported. Kubernetes and CRI are both currently agnostic about this format, passing the strings down from user-controlled APIs unchanged.

~~`DeviceMapping` no long accepts-and-ignores values in the `PathInContainer` or `CgroupPermissions` fields for Windows Devices.~~ (This will be a separate to-be-created PR, as this was a more-impactful change than I realised)

For reference, containerd has just [implemented the same syntax](https://github.com/containerd/containerd/pull/6618) for both CRI and the `ctr` CLI, to be part of the 1.7 release series. When [a more-specific CRI type is implemented for Windows Devices](https://github.com/kubernetes/kubernetes/issues/97739), then cri-dockerd (or other CRI implementation) will be responsible for packing the required data into this format for consumption by the daemon.

Windows Devices are no-longer limited to Process Isolation mode, as there are IDTypes that are also supported for HyperV Isolation, and I don't think it's desirable to repeat the mapping of which IDTypes work with which Isolation mode here.

**Note**: Due to limitations of the HCS v1 API used when not using the containerd backend:
* IDTypes other than `class` or `vpci-class-guid` are not supported without the containerd backend.
* Assigning devices to HyperV-isolated containers is not supported without the containerd backend.

Because these depend on the backend in-use, they are reported as failures from the backend via `ContainerStart` when not supported. Previously, `ContainerCreate` rejected both of these cases unconditionally, so these restrictions are looser than they were before this change.

Practically, these limitations from using the HCS v1 API won't matter, since Windows Server 2019 and earlier do not support the other IDType mappings, even if using the HCS v2 API via containerd, and Windows Server 2022 and later is intended to default to using the containerd backend and the HCS v2 API in Docker 21.xx. The only other Windows Server product not already end-of-life is SAC 20H2, [which does support the newer ID types](https://github.com/kubernetes/kubernetes/issues/97739#issuecomment-767981773). That product, the final Windows Server SAC release, reaches end-of-life in August 2022.

**- How I did it**

Changed the code that split `DeviceMapping{PathOnHost: "class/X"}` into `WindowsDevice{IDType:"class", ID: "X"}` to also split `DeviceMapping{PathOnHost: "W://X"}` into `WindowsDevice{IDType:"W", ID: "X"}`

**- How to verify it**

Unit test coverage for the parser.

Integration tests covering `class/5B45201D-F2F2-4F3B-85BB-30FF1F953599`, `class://5B45201D-F2F2-4F3B-85BB-30FF1F953599` and `vpci-class-guid://5B45201D-F2F2-4F3B-85BB-30FF1F953599` for both Process Isolation and HyperV isolation, supported by both containerd and HCS v1-based backends, where possible.

*Note*: Integration tests were having issues with HyperV isolation, so they have been skipped except for the early-failure case. It turns out we currently have no integration tests for HyperV isolation running on CI. This is possibly a nested-VM issue on the CI side, so it may vary by build-host config. I saw one pass for the skipped tests on an RS5 node.

Manually test exactly like the existing `--device` CLI flag on Windows, except using `class://` where you would currently use `class/`, e.g. `class/5B45201D-F2F2-4F3B-85BB-30FF1F953599` and `class://5B45201D-F2F2-4F3B-85BB-30FF1F953599` should produce identical results.

hcshim defines [a bunch of other possible IDTypes](https://github.com/microsoft/hcsshim/blob/v0.9.2/internal/uvm/virtual_device.go#L14-L21), but platform support and use-cases will vary, so I can't really advise on how to test them. And we don't deal with them here anyway, they go on opaquely down to hcshim to be handled.

**- Description for the changelog**

Windows Devices described by string, e.g. `--devices` in the CLI, now support the syntax `IDType://ID` as well as `class/ID`. Supported IDTypes depend on the active backend and other parameters of the container being created.

**- A picture of a cute animal (not mandatory but encouraged)**

[Baby Animal](https://muppet.fandom.com/wiki/Muppet_Babies_(puppets)?file=Character.babyanimal.jpg)

### CODE REVIEW

**Comment by TBBle** [comment]
I'm seeing a CI failure in the integration-cli test suite (I.e. after all my new tests have run), and have no idea how I could have introduced it.

```
[2022-03-13T11:16:32.522Z] === FAIL: github.com/docker/docker/integration-cli TestDockerSuite/TestPsListContainersFilterCreated (0.39s)
[2022-03-13T11:16:32.522Z]     docker_cli_ps_test.go:543: assertion failed: expression is false: strings.HasPrefix(cID, containerOut)
[2022-03-13T11:16:32.522Z]     --- FAIL: TestDockerSuite/TestPsListContainersFilterCreated (0.39s)
```

File: integration-cli/docker_cli_ps_test.go
Comment (TBBle) [inline integration-cli/docker_cli_ps_test.go:543]: This test was slightly faulty/weird... The `strings.HasPrefix` args were transposed, so it was checking if the `containerID` has the `ps -q` output as a prefix, which of course only works if the `ps -q` output lists *only* the container we want (in its truncated form). So a parallel run of another test could easily have broken it, for example.

This change is now more consistent with earlier in the same test, and also the other tests in this file which either use the same 12-character truncated ID, or call `docker ps -q --no-trunc` in order to compare the full container ID.

File: integration/container/devices_windows_test.go
Comment (TBBle) [inline integration/container/devices_windows_test.go:133]: As far as I can tell, we currently lack a general "does it work at all" integration test for HyperV-isolation on CI, where process isolation is the default, i.e. Windows Server.

I suspect the above-noted failures are caused by at least some of the test VMs not being configured for nested virtualisation, which is required for VM hosts to run HyperV-isolated containers, but not process-isolated containers.

As such, that probably needs to be investigated elsewhere (and remediated in the CI host setup if I'm correct about the root cause) before this `skip.If` can be removed.

I did mention above that there are HyperV tests, but they are actually tests that are only run when the daemon's default is HyperV, they're not testing HyperV specifically, i.e. [`TestRunHypervIsolationWithCPUCountCPUSharesAndCPUPercent`](https://github.com/moby/moby/blob/master/integration-cli/docker_cli_run_test.go#L4118-L4131).

File: libcontainerd/local/local_windows.go
Comment (TBBle) [inline libcontainerd/local/local_windows.go:318]: Although there are no current tests for this, I confirmed that it's hit with the existing tests before adding `vpci-class-guid` to be supported here. See da16170d50c5247ef618837cb29267fb731af8fb.

It'll take a bit more work to hit this code in a generic test, because every other IDType known to hcsshim is supported only in one of HyperV or Process isolation modes, ClassGUID is the only mechanism that works across both.

File: integration/container/devices_windows_test.go
Comment (thaJeztah) [inline integration/container/devices_windows_test.go:43]: I generally prefer being explicit for these, and to include the keys for each fields (it makes it easier to see what field is what, and doesn't require fields to be set to an empty value (if it's the default)

File: daemon/oci_windows_test.go
Comment (thaJeztah) [inline daemon/oci_windows_test.go:316]: More a nit / consistency thing; I don't think we use these kind of descriptions for (sub)tests, and generally use shorter terms (which may be easier to grep and/or use to run only specific sub-tests). I realise this convention is used by other projects and/or testing frameworks though 😅

**Review by thaJeztah — COMMENTED** [review]
Thanks! Generally, I think this change sounds good (I'd yet have to do a full review though). This change (indirectly) updates the API (as `container create` accepts the HostConfig struct). Can you add a mention in docs/api/version-history.md ?

Some thoughts I had while glancing over; most of this isn't _new_, and may require more substantial changes, so I would consider them really out of scope for this PR (but worth further discussion)

- The swagger documentation for the Devices struct is really under-documented [see this section](https://github.com/moby/moby/blob/efde293231f23d54719668e6c3643b5674a00ac1/api/swagger.yaml#L563-L567); that's already the case, so something to improve separately.
- The errors returned don't have an errdef-type, and looking at [`daemon.startContainer()`](https://github.com/moby/moby/blob/9674540ccff358c3cd84cc2f33c3503e0dab7fb7/daemon/start.go#L153-L156) it looks like they're converted to a `System` error (i.e. a `500` status on the API). This is also an existing issue (so for separate), and relates to the next point;
- It's unfortunate that we do all of this when creating the OCI Spec, which happens on container _start_. This means that a container can be created without error, even if fields are invalid, only to discover there's a problem when the container is _started_. While some of the validation / errors at "start" can't be avoided as they will occur when (e.g.) a device isn't there (anymore), and we also must perform checks when starting already existing containers, but ideally, we'd have handling for this on (before) container create, so that the user gets an error reported (400 invalid parameter) and no "faulty" container is created at all.
- To assist with the above, perhaps we should consider adding a `type=device` to the Mount API (`docker run --mount type=device,src=xxx,dest=xxx`); see https://github.com/moby/moby/issues/39293#issuecomment-666752243 (but I think there's other issues discussing that option as well).

File: daemon/oci_windows_test.go
Comment (TBBle) [reply daemon/oci_windows_test.go:316]: Umm... I copied this style from the other test in this file. Happy to shorten them though, they were a bit awkward to write as they are.

Comment (TBBle) [reply daemon/oci_windows_test.go:316]: Umm... I copied this style from the other test in this file. Happy to shorten them though, they were a bit awkward to write as they are.

**Comment by TBBle** [comment]
> Thanks! Generally, I think this change sounds good (I'd yet have to do a full review though). This change (indirectly) updates the API (as `container create` accepts the HostConfig struct). Can you add a mention in docs/api/version-history.md ?

Sure, will do. I'm assuming that 1.42 is API version for 21.xx, so I don't need to bump the API again. It seemed like the changelog here is more about structural changes than semantics, there's no mention of when `class/` was introduced as being snuck through the existing `PathOnHost` field. Looking again, I realise that the `Devices` array is never mentioned, so in hindsight I made a faulty assumption.

> Some thoughts I had while glancing over; most of this isn't _new_, and may require more substantial changes, so I would consider them really out of scope for this PR (but worth further discussion)
> 
> * The swagger documentation for the Devices struct is really under-documented [see this section](https://github.com/moby/moby/blob/efde293231f23d54719668e6c3643b5674a00ac1/api/swagger.yaml#L563-L567); that's already the case, so something to improve separately.

I did look at the Swagger docs, and they didn't seem to document the content or semantics of the `DeviceMapping` at all (particularly platform-specifics) so I left them alone. I recall needing to look up how to do multiple examples for a single field in OpenAPI 2, but don't recall if I actually found an answer. (I do know this changed in OpenAPI 3 with an `examples` field instead.).

> * The errors returned don't have an errdef-type, and looking at [`daemon.startContainer()`](https://github.com/moby/moby/blob/9674540ccff358c3cd84cc2f33c3503e0dab7fb7/daemon/start.go#L153-L156) it looks like they're converted to a `System` error (i.e. a `500` status on the API). This is also an existing issue (so for separate), and relates to the next point;
> * It's unfortunate that we do all of this when creating the OCI Spec, which happens on container _start_. This means that a container can be created without error, even if fields are invalid, only to discover there's a problem when the container is _started_. While some of the validation / errors at "start" can't be avoided as they will occur when (e.g.) a device isn't there (anymore), and we also must perform checks when starting already existing containers, but ideally, we'd have handling for this on (before) container create, so that the user gets an error reported (400 invalid parameter) and no "faulty" container is created at all.

Yeah, this was a big disappointment when I realised it was the case. It would be possible to add something to `daemon.containerCreate` to validate that the device syntax is correct, but the interesting errors of "valid syntax but IDType unknown to hcsshim" depend on calls that come later that `daemon.createSpec`, when turning OCI spec into HCS spec. hcsshim doesn't expose any API for validating these, AFAIK, particularly if you want to know if an IDType is not just known at all, but is valid in context of the rest of the spec.

> * To assist with the above, perhaps we should consider adding a `type=device` to the Mount API (`docker run --mount type=device,src=xxx,dest=xxx`); see [#39293 (comment)](https://github.com/moby/moby/issues/39293#issuecomment-666752243) (but I think there's other issues discussing that option as well).

That's an interesting idea, but I agree it's a bit out-of-scope for what I'm looking at here (because it's CLI usage, which was really just a nice bonus). It also doubles-down on the idea that for Windows devices, we're shoehorning device info via a string, since we're looking at `--mount type=device,src=class://GUID`, and that needs to end up with `dest=""`, not `dest=src` which might be the obvious default.

In the end, my main focus is to ensure that dockershim/cri-dockerd users aren't getting shut out of the new niceness I added to containerd. I don't know how valuable the IDTypes other than the already-functional `class` are to other Docker CLI/API users.

Because of this, it would be quite reasonable for example to say that for the Docker API, it should be `PathOnHost: IDType, PathInContainer: ID`, with an exception for existing `Path: class/GUID` behaviour (to cover current CLI and dockershim), and then it's up to the cri-dockerd code to parse the CRI HostPath  `://` into two fields instead of dockerd. That would also fit closer with the `--mount type=device` idea, particularly if the `src` and `dest` flags could take `idtype` and `id` as aliases, which is why I note this. (Long term, CRI should receive IDType and ID as two separate fields, so at that point cri-dockerd would be pasting them back into a single string for `PathOnHost`, which is odd but not unbearable.)

And [CDI](https://github.com/container-orchestrated-devices/container-device-interface) will make things slightly weird again, although perhaps dockerd won't deal with that, if it's implemented at the CRI layer. (Although its current spec'd as modding an OCI spec, so I guess it would _have_ to be supported in dockerd due to the way cri-dockerd is implemented? Or reimplement parts as modding a Docker API `HostConfig` instead.)

----

If the `://` separator itself is a concern, it'd be good to haggle that out soon, ideally aligning with containerd before it becomes a release build. Although I'm not really seeing any object to that aspect here, which is nice.

Note: I won't get a chance to get back to this code-wise until this coming weekend. I'm not seeing any complex or structure-changing changes to be made at this point, so I expect I can knock them out pretty easily given a couple of hours.

File: daemon/oci_windows_test.go
Comment (thaJeztah) [reply daemon/oci_windows_test.go:316]: oh! interesting; should've looked in this file itself ☺️. No worries; it was mostly a nit (and something that stood out, because we generally don't use this format, but apparently we did in this file); I'm fine keeping it as-is

**Comment by thaJeztah** [comment]
Quick reply to some of the above (have to jump out, so may add more later)

> Sure, will do. I'm assuming that 1.42 is API version for 21.xx, so I don't need to bump the API again

Correct; 1.42 is the "upcoming" version

> there's no mention of when class/ was introduced 

Yeah, I suspect there have been various changes that sneaked through in the past (some of them because they made changes "indirectly" through structs or parsing like the one that's being affected here).

> I did look at the Swagger docs, and they didn't seem to document the content or semantics of the DeviceMapping at all (particularly platform-specifics) so I left them alone.

That's ok; we still have a lot of things to address in the swagger file(s); I'm trying to tackle them one at a time (but there's many things under-documented).

**Comment by TBBle** [comment]
I've updated the API 1.42 version docs for the two changes I turned out to have made. I had a quick look, and confirmed that OpenAPI 2 does not support multiple examples per object definition, so the only way to show both syntaxes would be to have both appear in the same array in a higher-level example, which would produce a prima-facie unusable example.

Or upgrade to OpenAPI 3, which supports multiple examples for object definitions. That of-course is a much bigger and out-of-scope change.

File: integration/container/devices_windows_test.go
Comment (TBBle) [reply integration/container/devices_windows_test.go:105]: I logged this as #43395 so it can be looked at later.

File: docs/api/version-history.md
Comment (corhere) [inline docs/api/version-history.md:56]: It doesn't appear that the described behavior is aligned with the implementation. `func (*Daemon) createSpec` is only called in one place: `func (*Daemon) containerStart`. Therefore `PathInContainer` and `CgroupPermissions` will not be checked until one attempts to start the container.

Could this API change be versioned such that for API requests with version <= 1.41, the existing accept-but-ignore behavior is retained?

**Review by corhere — CHANGES_REQUESTED** [review]
The changes to the handling of the `PathInContainer` and `CgroupPermissions` fields on Windows containers is going to have unintended consequences. Everything else LGTM. Since that API change is tangential to introducing the `://` syntax for devices, would you consider spinning commit b63a0726ab70408407e242055f21ac350114da15 out into its own PR?

File: daemon/oci_windows.go
Comment (TBBle) [reply daemon/oci_windows.go:467]: Ah, I wasn't aware of that. Definitely happy to pull that change into into a separate PR, and perhaps rework it to deal with this correctly.

**Comment by TBBle** [comment]
I've broken out the `PathInContainer` and `CgroupPermissions` changes into a separate branch (https://github.com/TBBle/moby/tree/reject-non_empty-hostconfig-device-resources-we-cannot-use-on-windows) but I haven't created a PR yet since it's based on this one (adds to tests, changes a function created here). Once this lands, I'll rebase and create a draft PR to rework it based on @corhere's feedback.

That said, since it wasn't really a blocking change in any way (just seemed like low-hanging fruit, but not as low-hanging as I thought) it's not a high priority for me to advance that new PR right now.

File: integration-cli/docker_cli_ps_test.go
Comment (thaJeztah) [reply integration-cli/docker_cli_ps_test.go:543]: Thanks for fixing that; perhaps this was done to prevent any other string in the output (errors?) from causing a false "pass" (only guessing here). I expect that would only occur if the command _failed_, in which case the test would already be marked as failed, so I agree that this approach makes sense.

There's a lot of (some quite old) tests in `integration-cli` that could use some TLC some of them are quite brittle (and indeed running them in parallel, or in some cases, possible leftovers from previous tests, could make them fail).

**Review by thaJeztah — APPROVED** [review]
thank you! changes LGTM

unfortunately, looks like this needs one more rebase due to a conflict in the `api/changes.md` - happy to do a rebase for you if you prefer (should probably be easy to resolve)

File: libcontainerd/local/local_windows.go
Comment (TBBle) [reply libcontainerd/local/local_windows.go:317]: Yeah, I was thinking about doing exactly that earlier today. There's hcsshim/pkg/annotations, so something similar for these constants makes sense to me. I have a few (too many) hcsshim PRs in-flight though right now. -_- But this one's on my radar.

Comment (thaJeztah) [reply libcontainerd/local/local_windows.go:317]: Ah, yes, I know you have some other ones. Let's see if we can chat with some of the MS maintainers to get some moving.

BTW; are you on any of the docker or CNCF slack channels?

**Comment by TBBle** [comment]
* "ci / test-buildkit (./client, integration) (pull_request)" failure appears to have been a panic during a call to `t.Skipf`: Not related.
* Jenkins "win-2020-c8d" failed in "github.com/docker/docker/libnetwork/ipam TestRequestReleaseAddressDuplicate": Not related, but annoying because it didn't get to run the integration test suite. Also, given there's a `rand.Intn` in that test, the failure suggests that there's a subtle or uncommon bug in the test or the code under test.

**Comment by thaJeztah** [comment]
`TestRequestReleaseAddressDuplicate` is known to be flaky (quite some of the libnetwork tests we inherited from the libnetwork repository are known to be flaky (and need to be looked into / improved).

I'll kick CI once more 👍

File: libcontainerd/local/local_windows.go
Comment (TBBle) [reply libcontainerd/local/local_windows.go:317]: I'm not on any of the Slack channels, as it happens.

Comment (thaJeztah) [reply libcontainerd/local/local_windows.go:317]: Ah, that's a pity, thought it would be easier to (when needed) have a synchronous chat (also with the containerd maintainers, who mostly reside on the CNCF slack). If you ever decide to join that slack, let me know, then I can connect you with them as well 👍

**Comment by thaJeztah** [comment]
nice; CI is green now

@corhere PTAL

File: docs/api/version-history.md
Comment (TBBle) [inline docs/api/version-history.md:73]: Blast, looks like a rebasing error slipped in. I'll fix it shortly.
```suggestion
  depends on the underlying implementation and Windows version. This change is not versioned,
  and affects all API versions if the daemon has this patch.
```

File: integration/container/devices_windows_test.go
Comment (TBBle) [inline integration/container/devices_windows_test.go:104]:
```suggestion
			// TODO: Once https://github.com/moby/moby/issues/43395 is resolved,
			// remove this skip.If and validate the expected behaviour under Hyper-V.
```

File: integration/container/devices_windows_test.go
Comment (TBBle) [reply integration/container/devices_windows_test.go:105]: https://github.com/moby/moby/pull/43368/files#r835837598

File: libcontainerd/local/local_windows.go
Comment (TBBle) [reply libcontainerd/local/local_windows.go:318]: As part of [exposing and documenting these constants from hcsshim](https://github.com/moby/moby/pull/43368#discussion_r835767737), the coverage may be increased to cover this case (contrasting with the c8d backend not failing), although that is also increasing the amount of hcsshim-driven behaviour that is being tested.

**Comment by TBBle** [comment]
win-RS5 run timed out during integration-cli tests, everything else passed.

**Comment by thaJeztah** [comment]
good catch on the changelog rebase (I definitely missed it 😓); let me kick CI once more

**Review by tianon — APPROVED** [review]

**Pull request merged**
